test/e2e/e2e_test.go

```diff
@@ -125,3 +125,118 @@ func TestClusterHealthy(t *testing.T) {
 	// 'testEnv' is the env.Environment you set up in TestMain
 	_ = testEnv.Test(t, feature.Feature())
 }
+
+func TestScaling(t *testing.T) {
+	testCases := []struct {
+		name            string
+		initialSize     int
+		scaleTo         int
+		expectedMembers int
+	}{
+		{"ScaleDownFrom3To1", 3, 1, 1},
+		{"ScaleUpFrom1To3", 1, 3, 3},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			feature := features.New(tc.name)
+			etcdClusterName := fmt.Sprintf("etcd-%s", strings.ToLower(tc.name))
+
+			feature.Setup(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+				createEtcdCluster(ctx, t, c, etcdClusterName, tc.initialSize)
+				waitForStsReady(t, c, etcdClusterName, tc.initialSize)
+				return ctx
+			})
+
+			feature.Assess(
+				fmt.Sprintf("scale to %d", tc.scaleTo),
+				func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+					scaleEtcdCluster(ctx, t, c, etcdClusterName, tc.scaleTo)
+					return ctx
+				},
+			)
+
+			feature.Assess("verify member list", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+				podName := fmt.Sprintf("%s-0", etcdClusterName)
+				stdout, stderr, err := execInPod(t, c, podName, namespace, []string{"etcdctl", "member", "list"})
+				if err != nil {
+					t.Fatalf("Failed to exec in pod: %v, stderr: %s", err, stderr)
+				}
+
+				if len(strings.Split(strings.TrimSpace(stdout), "\n")) != tc.expectedMembers {
+					t.Errorf("Expected to find %d members in member list, but got: %s", tc.expectedMembers, stdout)
+				}
+				return ctx
+			})
+
+			feature.Teardown(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+				cleanupEtcdCluster(ctx, t, c, etcdClusterName)
+				return ctx
+			})
+
+			_ = testEnv.Test(t, feature.Feature())
+		})
+	}
+}
+
+func TestPromoteReadyLearner(t *testing.T) {
+	feature := features.New("promote-learner")
+	etcdClusterName := "etcd-promote-learner"
+
+	feature.Setup(func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+		createEtcdCluster(ctx, t, c, etcdClusterName, 2)
+		waitForStsReady(t, c, etcdClusterName, 2)
+		// Manually add a third member as a learner
+		podName := fmt.Sprintf("%s-0", etcdClusterName)
+		command := []string{
+			"etcdctl", "member", "add", "etcd-promote-learner-2",
+			"--peer-urls=http://etcd-promote-learner-2.etcd-promote-learner.etcd-operator-system.svc.cluster.local:2380",
+			"--learner",
+		}
```

Comment:
If TLS is enabled, the command will fail. More importantly, why adding member directly? It should be the responsibility of the controller.

Reply:
I agree, maybe we should keep the test case in unit test rather than make it E2E.https://github.com/etcd-io/etcd-operator/blob/267dbb96e08c332cecca3bd698d64d07c7e6ac55/internal/etcdutils/etcdutils_test.go#L122
Removed.